### ISSUE TITLE
Spurious exhaustivity warning in switch expression on non-nullable property

### ISSUE THREAD

**Issue by 333fred** [dotnet/roslyn#53626]
We appear to have a bug around pure null checks being _omitted_ from code:

```cs
#nullable enable
public class C {
    public string Prop { get; set; } = "";
    public bool B { get; set; }

    public void M1() {
        _ = this switch // No warning
        {
            { B: true, Prop: { } e } => "",
            { Prop: { } e } => "",
        };
    }
    public void M2() {
        _ = this switch // No warning
        {
            { B: true, Prop: { } e } => "",
            _ => "",
        };
    }
    public void M3() {
        _ = this switch // Warning: non-exhaustive, doesn't handle Prop being null
        {
            { B: true, Prop: { } e } => "",
            { } => "",
        };
    }
}
```

`M3` makes a pure null check on `this` (which is never false, of course), but that check is carrying through to its nested slots, which seems incorrect.

/cc @RikkiGibson if there's something I've misunderstood here.

https://sharplab.io/#v2:EYLgtghglgdgPgYhgVwDaosVBTABNmTHAWACgABAZl3ICZcBhXAbzN3ZuvIEYAGXAAoAnAPYAHFrgDm2AC4BuXAGc5igL64AvLgBEO+Ww5VcwESNS4AQpJkLlq3GrKH2x8gBZcAWW4AKAJQsLhzsAPpauLIAFlBKygDuULIAxlG4APTpuAByIrjxEEIwsFLBIaykIVUczFYgkULI2AA0gqJi9bUaeBqaAHy6Os1l1ZLC4p2O+FP9g8OVo2oGC+xOK5w0nl60AUHrIeHa0bEJSakZWbn5hcUwpfs1I9W1lvWyjS1tE5LdMwN681GVUO/yGT0cyxCaxCbi2lF2FSBh0iMTiSkSKTSmVwAHUbiV6jARDAALTYAAeUQgyCUsigADdPgATETYJQwADkslwVJgTJwXwkwGwJVwKHQ4MRQJqdQaTVa4w6P2mvVBgOl7C6WjV4KWwTWaiAA=

**Comment by RikkiGibson** [comment]
It looks like the diagnostic expects you to handle the case where `this` is `null` because the empty `{ }` is treated as a pure null test.

Equivalent warning is given in `M4` in the linked scenario: [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgPgYhgVwDaosVBTABNmTHAWACgABAZl3ICZcBhXAbzN3ZuvIEYAGXAAoAnAPYAHFrgDm2AC4BuXAGc5igL64AvLgBEO+Ww5VcwESNS4AQpJkLlq3GrKH2x8gBZcAWW4AKAJQsLhzsAPpauLIAFlBKygDuULIAxlG4APTpuAByIrjxEEIwsFLBIaykIVUczFYgkULI2AA0gqJi9bUaeBqaAHy6Os1l1ZLC4p2O+FP9g8OVo2oGC+xOK5w0nl60AUHrIeHa0bEJSakZWbn5hcUwpfs1I9W1lvWyjS1tE5LdMwN681GVUO/yGT0cyxCaxCbi2lF2FSBh0iMTiSkSKTSmVwAHUbiV6jARDAALTYAAeUQgyCUsigADdPgATETYJQwADkslwVJgTJwXwkwGwJVwKHQ4MRQJqdQaTVa4w6P2mvVBgOl7C6WjV4KWwWhRi4W3cCPByOOaIx52xeKKBLFxLJlOptIZzNZ7K5PIgfIFXNRHJMIruYrQqEl4PKfzmushHDWaiAA=)

**Comment by 333fred** [comment]
Hmm, I think I was misreading the error message. Nevermind, this is correct.
